integrations/dav-server/src/opendalfs.rs

```diff
@@ -0,0 +1,216 @@
+            let exist = self.op.is_exist(path).await;
+            match exist {
+                Ok(exist) => match exist {
+                    true => Err(dav_server::fs::FsError::Exists),
+                    false => {
+                        let res = self.op.create_dir(path).await;
+                        match res {
+                            Ok(_) => Ok(()),
+                            Err(e) => Err(convert_error(e)),
+                        }
+                    }
+                },
+                Err(e) => Err(convert_error(e)),
+            }
+        }
+        .boxed()
+    }
+
+    fn remove_file<'a>(&'a self, path: &'a DavPath) -> dav_server::fs::FsFuture<()> {
+        async move {
+            self.op
+                .delete(path.as_url_string().as_str())
+                .await
+                .map_err(convert_error)
+        }
+        .boxed()
+    }
+
+    fn remove_dir<'a>(&'a self, path: &'a DavPath) -> dav_server::fs::FsFuture<()> {
+        self.remove_file(path)
+    }
+
+    fn copy<'a>(&'a self, from: &'a DavPath, to: &'a DavPath) -> dav_server::fs::FsFuture<()> {
+        async move {
+            let from_path = from
+                .as_rel_ospath()
+                .to_str()
+                .ok_or(FsError::GeneralFailure)?;
+            let to_path = to.as_rel_ospath().to_str().ok_or(FsError::GeneralFailure)?;
+            self.op
+                .copy(from_path, to_path)
+                .await
+                .map_err(convert_error)
+        }
+        .boxed()
+    }
+
+    fn rename<'a>(&'a self, from: &'a DavPath, to: &'a DavPath) -> dav_server::fs::FsFuture<()> {
+        async move {
+            let from_path = from
+                .as_rel_ospath()
+                .to_str()
+                .ok_or(FsError::GeneralFailure)?;
+            let to_path = to.as_rel_ospath().to_str().ok_or(FsError::GeneralFailure)?;
+            if from.is_collection() {
+                let _ = self.remove_file(to).await;
+            }
+            self.op
+                .rename(from_path, to_path)
+                .await
+                .map_err(convert_error)
+        }
+        .boxed()
+    }
+}
+
+impl OpendalFs {
+    pub fn new(op: Operator) -> Box<OpendalFs> {
+        Box::new(OpendalFs { op })
+    }
+}
+
+struct DavStream {
+    op: Operator,
+    lister: Lister,
+}
+
+impl Stream for DavStream {
+    type Item = Box<dyn DavDirEntry>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let dav_stream = self.get_mut();
+        let lister = Pin::new(&mut dav_stream.lister).get_mut();
+
+        match Pin::new(lister).poll_next(cx) {
+            Ready(entry) => match entry {
+                Some(entry) => {
+                    let webdav_entry = WebDAVDirEntry::new(entry.unwrap(), dav_stream.op.clone());
```

Comment:
Please don't unwrap error returned by OpenDAL, return en error instead.

Since `FsStream` requires to return `Box<dyn DavDirEntry>`, maybe we return a bad entry instead.

This change can be done in following PRs.

Reply:
sorry I don't get this. how to make a bad entry? `Entry::new(path, metadata)` is marked as `pub(crate)`.

Reply:
Hi, @messense, how should we handle errors that occur when reading a directory? Do you have any advice for us? Thanks in advance.